File: cpp/16-string-compression/src/solution.cpp

#include <sstream>
#include <string>
#include "solution.h"

std::string Compress(std::string s) {
    std::ostringstream compressed;
    int counter = 0;
    char prevChar = *s.begin();

    for (std::string::iterator c = s.begin(); c != s.end(); ++c) {
        if (prevChar != *c && counter != 0) {
            compressed << prevChar << std::to_string(counter);
            prevChar = *c;
            counter = 1;
        } else if (prevChar == *c) {
            counter += 1;
        }
    }
    compressed << prevChar << std::to_string(counter);

    std::string compressedString = compressed.str();

    return
        s.length() > compressedString.length()
            ? compressedString
            : s;
}

File: py/16-string-compression/solution.py

def compress(s):
	compressed = ''
	prev_letter, prev_letter_count = '', 0

	for i, c in enumerate(s):
		if c == prev_letter:
			prev_letter_count += 1
		elif prev_letter_count > 0:
			compressed += prev_letter + str(prev_letter_count)
			prev_letter, prev_letter_count = c, 1
		else:
			prev_letter, prev_letter_count = c, 1

	if prev_letter_count > 0:
		compressed += prev_letter + str(prev_letter_count)

	return compressed
